rpc/transport/ws/server.go

```diff
@@ -71,18 +72,60 @@ func (wsc *serverWebSocketTransport) Close() {
 }
 
 func (wsc *serverWebSocketTransport) Url() string {
-	return "ws://" + webscocketServerAddress + wsc.port
+	return webscocketServerAddress + wsc.port
+}
+
+func (wsc *serverWebSocketTransport) request(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+		return
+	}
+	body := http.MaxBytesReader(w, r.Body, 8192)
+	msg, err := io.ReadAll(body)
+	if err != nil {
+		http.Error(w, http.StatusText(http.StatusRequestEntityTooLarge), http.StatusRequestEntityTooLarge)
+		return
+	}
+
+	_, err = w.Write(wsc.requestHandler(msg))
+	if err != nil {
+		panic(err)
+	}
 }
 
-func (wsc *serverWebSocketTransport) subscribeRequestHandler(w http.ResponseWriter, r *http.Request) {
+func (wsc *serverWebSocketTransport) subscribe(w http.ResponseWriter, r *http.Request) {
 	c, err := websocket.Accept(w, r, nil)
 	if err != nil {
 		panic(err)
 	}
-	wsc.serverWebsocket = c
 	defer c.Close(websocket.StatusInternalError, "server initiated websocket close")
+	notificationChan := make(chan []byte)
+	key := strconv.Itoa(int(rand.Uint64()))
+	wsc.notificationListeners.Store(key, notificationChan)
+	defer wsc.notificationListeners.Delete(key)
+
+	// A client closes a connection by sending a message over the websocket
+	closeChan := make(chan error)
+	go func() {
+		_, _, err := c.Read(r.Context())
+		closeChan <- err
+	}()
 
-	err = wsc.readRequests(r.Context())
+	done := false
+	for !done {
+		select {
+		case err = <-closeChan:
+			done = true
+		case <-r.Context().Done():
+			err = r.Context().Err()
+			done = true
+		case notificationData := <-notificationChan:
+			err := c.Write(r.Context(), websocket.MessageText, notificationData)
+			if err != nil {
+				done = true
```

Comment:
Outside the scope of this PR but we could consider using a `break` [statement with a label](https://go.dev/ref/spec#Break_statements).  I think the explicit behavior of `break` reduces some mental overhead of making sure nothing executes after `done=true` is set.

Reply:
Good idea! I wasn't aware of labels. af9e49a8